src/order-lazy.h

```diff
@@ -0,0 +1,141 @@
+#ifndef VCTRS_ORDER_LAZY_H
+#define VCTRS_ORDER_LAZY_H
+
+#include "vctrs.h"
+
+// -----------------------------------------------------------------------------
+
+/*
+ * `lazy_vec` is a lazy raw vector that only allocates itself when
+ * `lazy_vec_initialize()` is called. It is used as working memory of varying
+ * types by `vec_order()`. In `vec_order()` we aren't always sure how much
+ * working memory is required, but we want to reuse it where we can once we
+ * do allocate it. These lazy vectors allow us to specify the maximum amount
+ * required up front without actually allocating them, delaying that allocation
+ * until they are truly needed.
+ *
+ * @member data The RAWSXP that gets allocated lazily.
+ * @member p_data A void pointer to the RAWSXP.
+ * @member data_pi A protection index to `data` so it can reprotect itself
+ *   upon allocation.
+ * @member size The total size of the RAWSXP to allocate. This is generally
+ *   computed as `size * multiplier` in `new_lazy_vec()`, where multiplier
+ *   is a `sizeof(<type>)`.
+ * @member initialized Has the lazy vector been initialized yet?
+ */
+struct lazy_vec {
```

Comment:
Hmm, I thought it was a little odd to have arguments like `struct lazy_raw* p_lazy_x_aux` since the intention of `p_lazy_x_aux` is not to be a raw vector. It was just an implementation detail that it was constructed that way

Reply:
I think the code will be simpler to understand if you use independent constructs that can be understood in isolation. Otherwise the reader feels like they need to understand both the main code and the helpers to get a complete view. Also, the isolated purpose of these helpers make it easier to reuse later and these helpers seem generally useful. (No need to go out of your way to apply this principle though, as it would be easy to overengineer.)

In that context, `lazy_raw` seems a better name since `raw` conveys the allocation of a C array on the R heap.